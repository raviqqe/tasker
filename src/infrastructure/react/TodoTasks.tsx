import { PulseLoader } from "react-spinners";
import React from "react";
import { SortableContainer, SortableElement } from "react-sortable-hoc";
import styled from "styled-components";
import { ITask } from "../../domain/task";
import { Task } from "./Task";
import { buttonMargin } from "./style";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  overflow: auto;
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Tasks = SortableContainer(styled.div`
  display: flex;
  flex-direction: column;
  padding: 1em 0.5em ${buttonMargin};

  > * {
    margin: 0.5em;
  }
`);

const SortableTask = SortableElement(Task);

export interface IProps {
  completeTodoTask: (task: ITask) => Promise<void>;
  reorderTodoTasks: (taskIDs: string[]) => Promise<void>;
  todoTasks: ITask[] | null;
  updateTodoTask: (task: ITask) => Promise<void>;
}

export const TodoTasks = ({
  completeTodoTask,
  reorderTodoTasks,
  todoTasks,
  updateTodoTask
}: IProps) =>
  todoTasks ? (
    <Container>
      <Tasks
        distance={4}
        onSortEnd={async ({ oldIndex, newIndex }) => {
          const taskIDs = todoTasks.map(task => task.id);
          taskIDs.splice(newIndex, 0, taskIDs.splice(oldIndex, 1)[0]);
          await reorderTodoTasks(taskIDs);
        }}
        useDragHandle={true}
      >
        {todoTasks.map((task: ITask, index: number) => (
          <SortableTask
            completeTask={completeTodoTask}
            dragHandleEnabled={true}
            index={index}
            key={task.id}
            task={task}
            updateTask={updateTodoTask}
          />
        ))}
      </Tasks>
    </Container>
  ) : (
    <LoaderContainer>
      <PulseLoader color="white" />
    </LoaderContainer>
  );
