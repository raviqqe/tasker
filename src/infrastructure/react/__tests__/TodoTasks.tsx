import React from "react";
import { create } from "react-test-renderer";
import { TodoTasks } from "../TodoTasks";

it("renders", () => {
  expect(
    create(
      <TodoTasks
        completeTodoTask={async () => {}}
        reorderTodoTasks={async () => {}}
        todoTasks={[{ id: "id", name: "name" }]}
        updateTodoTask={async () => {}}
      />
    ).toJSON()
  ).toMatchSnapshot();
});

it("renders with no tasks", () => {
  expect(
    create(
      <TodoTasks
        completeTodoTask={async () => {}}
        reorderTodoTasks={async () => {}}
        todoTasks={[]}
        updateTodoTask={async () => {}}
      />
    ).toJSON()
  ).toMatchSnapshot();
});

it("renders with tasks not loaded yet", () => {
  expect(
    create(
      <TodoTasks
        completeTodoTask={async () => {}}
        reorderTodoTasks={async () => {}}
        todoTasks={null}
        updateTodoTask={async () => {}}
      />
    ).toJSON()
  ).toMatchSnapshot();
});
