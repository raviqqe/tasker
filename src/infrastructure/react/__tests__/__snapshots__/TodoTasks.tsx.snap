// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="sc-jSMfEi cqVjBh"
>
  <div
    class="sc-iBkjds tkAkl"
  >
    <div
      aria-describedby="DndDescribedBy-0"
      aria-disabled="false"
      aria-roledescription="sortable"
      class="sc-gsnTZi GZWTk"
      role="button"
      tabindex="0"
    >
      <div
        class="sc-dkzDqf grNauI"
      >
        name
      </div>
      <div
        class="sc-hKMtZM ctIavJ"
      >
        <div
          aria-label="Done"
          class="sc-bczRLJ bVpJri"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M9 16.17L4.83 12l-1.42 1.41L9 19 21 7l-1.41-1.41z"
            />
          </svg>
        </div>
        <div
          aria-label="Edit"
          class="sc-bczRLJ bVpJri"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
            />
          </svg>
        </div>
        <div
          class="sc-bczRLJ bVpJri sc-eCYdqJ bKyqnI"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders with no tasks 1`] = `
<div
  class="sc-jSMfEi cqVjBh"
>
  <div
    class="sc-iBkjds tkAkl"
  />
</div>
`;

exports[`renders with tasks not loaded yet 1`] = `
<div
  class="sc-gKXOVf kQgVQy"
>
  <span
    class="css-1xdhyk6"
  >
    <span
      class="css-hqq8du"
    />
    <span
      class="css-1h0qtnz"
    />
    <span
      class="css-1ihj5oh"
    />
  </span>
</div>
`;
